src/Sarif.Viewer.VisualStudio/CodeAnalysisResultManager.cs

```diff
@@ -239,24 +243,89 @@ public bool TryRebaselineCurrentSarifError(string uriBaseId, string originalFile
 
             string rebaselinedFile;
 
-            if (Uri.IsWellFormedUriString(originalFilename, UriKind.Absolute))
+            if (FileDetails.ContainsKey(originalFilename))
             {
-                rebaselinedFile = DownloadFile(originalFilename);
+                // File contents embedded in SARIF.
+                rebaselinedFile = CreateFileFromContents(originalFilename);
             }
             else
             {
-                rebaselinedFile = GetRebaselinedFileName(uriBaseId, originalFilename);
-            }
+                if (Uri.IsWellFormedUriString(originalFilename, UriKind.Absolute))
+                {
+                    // File needs to be downloaded.
+                    rebaselinedFile = DownloadFile(originalFilename);
+                }
+                else
+                {
+                    // User needs to locate file.
+                    rebaselinedFile = GetRebaselinedFileName(uriBaseId, originalFilename);
+                }
 
-            if (String.IsNullOrEmpty(rebaselinedFile) || originalFilename.Equals(rebaselinedFile, StringComparison.OrdinalIgnoreCase))
-            {
-                return false;
+                if (String.IsNullOrEmpty(rebaselinedFile) || originalFilename.Equals(rebaselinedFile, StringComparison.OrdinalIgnoreCase))
+                {
+                    return false;
+                }
             }
 
             CurrentSarifError.RemapFilePath(originalFilename, rebaselinedFile);
             return true;
         }
 
+        // Contents are embedded in SARIF. Create a file from these contents.
+        private string CreateFileFromContents(string fileName)
+        {
+            var fileData = FileDetails[fileName];
+
+            // Get temporary path.
+            string path = Path.GetTempPath();
+            path = Path.Combine(path, TemporaryFileDirectoryName);
+
+            // If the file path already starts with the temporary location,
+            // that means we've already built the temporary file, so we can
+            // just open it.
+            if (fileName.StartsWith(path))
+            {
+                path = fileName;
+            }
+            // Else we have to create a location under the temp path.
+            else
+            {
+                // If the file path is absolute, or if it is relative to an
+                // unknown root, we need to strip the drive letter and/or
+                // leading slash, so that we can relate the file path to an
+                // arbitrary remapped disk location.
+                string filePath = fileName;
+                filePath = filePath.Substring(Path.GetPathRoot(filePath).Length);
+
+                path = Path.Combine(path, fileData.Sha256Hash);
+                path = Path.Combine(path, filePath);
```

Comment:
"Good news, everyone!" Since .NET 4, there's a 3-argument overload:
```
path = Path.Combine(path, fileData.Sha256Hash, filePath);
```

Also: worth a comment explaining how you know `Sha256Hash` is guaranteed to be present.

Reply:
That's one of our requirements. For now, you can't show an embedded file that doesn't have a hash (and as per Michael's thoughts, in the future we'll calculate the hash if it's not present).